jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -0,0 +1,711 @@
+#include "ecma-objects.h"
+#include "ecma-objects-general.h"
+#include "ecma-promise-object.h"
+
+#ifndef CONFIG_DISABLE_ES2015_PROMISE_BUILTIN
+
+/** \addtogroup ecma ECMA
+ * @{
+ *
+ * \addtogroup ecmapromiseobject ECMA Promise object related routines
+ * @{
+ */
+
+/**
+ * Check if an object is promise.
+ *
+ * @return true - if it is a promise object.
+ *         false - if it is not a promise object.
+ */
+inline bool __attr_always_inline___
+ecma_is_promise (ecma_object_t *obj_p) /**< points to object */
+{
+  return ecma_object_class_is (obj_p, LIT_MAGIC_STRING_PROMISE_UL);
+} /* ecma_is_promise */
+
+/**
+ * Get the result of the promise.
+ *
+ * @return ecma value of the promise result.
+ *         Returned value must be freed with ecma_free_value
+ */
+inline ecma_value_t
+ecma_promise_get_result (ecma_object_t *obj_p) /**< points to promise object */
+{
+  JERRY_ASSERT (ecma_is_promise (obj_p));
+
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) obj_p;
+
+  return ecma_copy_value (ext_object_p->u.class_prop.u.value);
+} /* ecma_promise_get_result */
+
+/**
+ * Set the PromiseResult of promise.
+ */
+inline void __attr_always_inline___
+ecma_promise_set_result (ecma_object_t *obj_p, /**< points to promise object */
+                         ecma_value_t result) /**< the result value */
+{
+  JERRY_ASSERT (ecma_is_promise (obj_p));
+
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) obj_p;
+
+  JERRY_ASSERT (ext_object_p->u.class_prop.u.value == ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED));
+
+  ext_object_p->u.class_prop.u.value = result;
+} /* ecma_promise_set_result */
+
+/**
+ * Get the PromiseState of promise.
+ *
+ * @return the state's enum value
+ */
+inline uint8_t __attr_always_inline___
+ecma_promise_get_state (ecma_object_t *obj_p) /**< points to promise object */
+{
+  JERRY_ASSERT (ecma_is_promise (obj_p));
+
+  return ((ecma_promise_object_t *) obj_p)->state;
+} /* ecma_promise_get_state */
+
+/**
+ * Set the PromiseState of promise.
+ */
+inline void __attr_always_inline___
+ecma_promise_set_state (ecma_object_t *obj_p, /**< points to promise object */
+                        uint8_t state) /**< the state */
+{
+  JERRY_ASSERT (ecma_is_promise (obj_p));
+
+  ((ecma_promise_object_t *) obj_p)->state = state;
+} /* ecma_promise_set_state */
+
+/**
+ * Get the bool value of alreadyResolved.
+ *
+ * @return bool value of alreadyResolved.
+ */
+static bool
+ecma_get_already_resolved_bool_value (ecma_value_t already_resolved) /**< the alreadyResolved */
+{
+  JERRY_ASSERT (ecma_is_value_object (already_resolved));
+
+  ecma_object_t *already_resolved_p = ecma_get_object_from_value (already_resolved);
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) already_resolved_p;
+
+  JERRY_ASSERT (ext_object_p->u.class_prop.class_id == LIT_MAGIC_STRING_BOOLEAN_UL);
+
+  return ext_object_p->u.class_prop.u.value == ecma_make_boolean_value (true);
```

Comment:
because both promise's reject_function and resolve_function share the same "alreadyResolved" value. (the boolean object). so that the reject or resolve function will know whether the promise is processed by checking the "alreadyResolved".